File: crates/chthl/src/lexer.rs
//! This module defines the Lexer, which is responsible for turning CHTL source code
//! into a stream of tokens.

use crate::token::Token;

pub struct Lexer<'a> {
    input: &'a [u8],
    position: usize,
    read_position: usize,
    ch: u8,
}

impl<'a> Lexer<'a> {
    pub fn new(input: &'a str) -> Self {
        let mut l = Lexer {
            input: input.as_bytes(),
            position: 0,
            read_position: 0,
            ch: 0,
        };
        l.read_char();
        l
    }

    fn read_char(&mut self) {
        if self.read_position >= self.input.len() {
            self.ch = 0;
        } else {
            self.ch = self.input[self.read_position];
        }
        self.position = self.read_position;
        self.read_position += 1;
    }

    fn peek_char(&self) -> u8 {
        if self.read_position >= self.input.len() {
            0
        } else {
            self.input[self.read_position]
        }
    }

    fn skip_whitespace(&mut self) {
        while self.ch.is_ascii_whitespace() {
            self.read_char();
        }
    }

    pub fn next_token(&mut self) -> Token {
        self.skip_whitespace();

        let tok = match self.ch {
            b'=' => { self.read_char(); Token::Assign }
            b';' => { self.read_char(); Token::Semicolon }
            b'(' => { self.read_char(); Token::LParen }
            b')' => { self.read_char(); Token::RParen }
            b',' => { self.read_char(); Token::Comma }
            b'{' => { self.read_char(); Token::LBrace }
            b'}' => { self.read_char(); Token::RBrace }
            b'[' => { self.read_char(); Token::LBracket }
            b']' => { self.read_char(); Token::RBracket }
            b':' => { self.read_char(); Token::Colon }
            b'!' => { self.read_char(); Token::Bang }
            b'+' => { self.read_char(); Token::Plus }
            b'*' => { self.read_char(); Token::Asterisk }
            b'@' => { self.read_char(); Token::At }
            b'#' => { self.read_char(); Token::Hash }
            b'.' => { self.read_char(); Token::Dot }
            b'?' => { self.read_char(); Token::Question }
            b'>' => { self.read_char(); Token::Gt }
            b'<' => { self.read_char(); Token::Lt }
            b'&' if self.peek_char() == b'&' => { self.read_char(); self.read_char(); Token::LogicalAnd }
            b'|' if self.peek_char() == b'|' => { self.read_char(); self.read_char(); Token::LogicalOr }
            b'-' if self.peek_char() == b'>' => { self.read_char(); self.read_char(); Token::Arrow }
            b'-' if self.peek_char() == b'-' => { self.read_char(); self.read_char(); Token::GeneratorComment(self.read_comment_line()) }
            b'-' => { self.read_char(); Token::Minus }
            b'/' if self.peek_char() == b'/' => { self.read_comment_line(); self.next_token() }
            b'/' if self.peek_char() == b'*' => { self.read_multiline_comment(); self.next_token() }
            b'/' => { self.read_char(); Token::Slash }
            b'&' => { self.read_char(); Token::Ampersand }
            b'"' | b'\'' => {
                let quote = self.ch;
                let content = self.read_string(quote);
                self.read_char();
                Token::String(content)
            }
            0 => Token::Eof,
            _ => {
                if self.ch.is_ascii_digit() {
                    return Token::Number(self.read_number());
                }
                if is_identifier_start(self.ch) {
                    return Token::Ident(self.read_identifier());
                }
                let ch = self.ch;
                self.read_char();
                Token::Illegal(String::from_utf8_lossy(&[ch]).to_string())
            }
        };
        tok
    }

    fn read_identifier(&mut self) -> String {
        let position = self.position;
        while is_identifier_char(self.ch) {
            self.read_char();
        }
        String::from_utf8_lossy(&self.input[position..self.position]).to_string()
    }

    fn read_number(&mut self) -> String {
        let position = self.position;
        while self.ch.is_ascii_digit() {
            self.read_char();
        }
        if self.ch == b'.' {
            self.read_char();
            while self.ch.is_ascii_digit() {
                self.read_char();
            }
        }
        String::from_utf8_lossy(&self.input[position..self.position]).to_string()
    }

    fn read_string(&mut self, quote: u8) -> String {
        let position = self.position + 1;
        loop {
            self.read_char();
            if self.ch == quote || self.ch == 0 {
                break;
            }
        }
        String::from_utf8_lossy(&self.input[position..self.position]).to_string()
    }

    fn read_comment_line(&mut self) -> String {
        let position = self.position;
        while self.ch != b'\n' && self.ch != 0 {
            self.read_char();
        }
        String::from_utf8_lossy(&self.input[position..self.position]).trim().to_string()
    }

    fn read_multiline_comment(&mut self) {
        self.read_char();
        self.read_char();
        loop {
            if self.ch == b'*' && self.peek_char() == b'/' {
                self.read_char();
                self.read_char();
                break;
            }
            if self.ch == 0 {
                break;
            }
            self.read_char();
        }
    }
}

fn is_identifier_start(ch: u8) -> bool {
    ch.is_ascii_alphabetic() || ch == b'_'
}

fn is_identifier_char(ch: u8) -> bool {
    ch.is_ascii_alphanumeric() || ch == b'_' || ch == b'-'
}

#[cfg(test)]
mod tests {
    use super::*;
    use crate::token::Token;

    #[test]
    fn test_next_token_simple() {
        let input = "div { id: box; }";
        let tests = vec![
            Token::Ident("div".to_string()),
            Token::LBrace,
            Token::Ident("id".to_string()),
            Token::Colon,
            Token::Ident("box".to_string()),
            Token::Semicolon,
            Token::RBrace,
            Token::Eof,
        ];
        let mut l = Lexer::new(input);
        for expected_token in tests {
            assert_eq!(l.next_token(), expected_token);
        }
    }

    #[test]
    fn test_css_value_as_identifier() {
        let input = "width: 100px;";
        let tests = vec![
            Token::Ident("width".to_string()),
            Token::Colon,
            Token::Number("100".to_string()),
            Token::Ident("px".to_string()),
            Token::Semicolon,
            Token::Eof,
        ];
        let mut l = Lexer::new(input);
        for expected_token in tests {
            assert_eq!(l.next_token(), expected_token);
        }
    }

    #[test]
    fn test_comments() {
        let input = r#"
            // This is a single line comment
            div {
                -- This is a generator comment
                /* This is a
                   multiline comment */
                color: red;
            }
        "#;
        let tests = vec![
            Token::Ident("div".to_string()),
            Token::LBrace,
            Token::GeneratorComment("This is a generator comment".to_string()),
            Token::Ident("color".to_string()),
            Token::Colon,
            Token::Ident("red".to_string()),
            Token::Semicolon,
            Token::RBrace,
            Token::Eof,
        ];
        let mut l = Lexer::new(input);
        for expected_token in tests {
            assert_eq!(l.next_token(), expected_token);
        }
    }

    #[test]
    fn test_all_symbols() {
        let input = r#"=;(),{}[];:@#.>&&||->?"#;
        let tests = vec![
            Token::Assign,
            Token::Semicolon,
            Token::LParen,
            Token::RParen,
            Token::Comma,
            Token::LBrace,
            Token::RBrace,
            Token::LBracket,
            Token::RBracket,
            Token::Semicolon,
            Token::Colon,
            Token::At,
            Token::Hash,
            Token::Dot,
            Token::Gt,
            Token::LogicalAnd,
            Token::LogicalOr,
            Token::Arrow,
            Token::Question,
            Token::Eof,
        ];
        let mut l = Lexer::new(input);
        for expected_token in tests {
            assert_eq!(l.next_token(), expected_token, "Failed on input: {}", input);
        }
    }

    #[test]
    fn test_numbers_and_operators() {
        let input = "10 + 20.5 * 3 / 4 - 5 !";
        let tests = vec![
            Token::Number("10".to_string()),
            Token::Plus,
            Token::Number("20.5".to_string()),
            Token::Asterisk,
            Token::Number("3".to_string()),
            Token::Slash,
            Token::Number("4".to_string()),
            Token::Minus,
            Token::Number("5".to_string()),
            Token::Bang,
            Token::Eof,
        ];
        let mut l = Lexer::new(input);
        for expected_token in tests {
            assert_eq!(l.next_token(), expected_token);
        }
    }
}

File: pkg/lexer/lexer.go
package lexer

type Lexer struct {
	input        string
	position     int
	readPosition int
	ch           byte
}

func New(input string) *Lexer {
	l := &Lexer{input: input}
	l.readChar()
	return l
}

func (l *Lexer) readChar() {
	if l.readPosition >= len(l.input) {
		l.ch = 0
	} else {
		l.ch = l.input[l.readPosition]
	}
	l.position = l.readPosition
	l.readPosition++
}

func (l *Lexer) peekChar() byte {
	if l.readPosition >= len(l.input) {
		return 0
	}
	return l.input[l.readPosition]
}

func (l *Lexer) NextToken() Token {
	var tok Token
	l.skipWhitespace()

	switch l.ch {
	case '=': tok = newToken(ASSIGN, l.ch)
	case ':': tok = newToken(COLON, l.ch)
	case ';': tok = newToken(SEMICOLON, l.ch)
	case '+': tok = newToken(PLUS, l.ch)
	case '-':
		if l.peekChar() == '-' {
			l.readChar(); l.readChar()
			tok.Type = GEN_COMMENT
			tok.Literal = l.readLineComment()
			return tok
		}
		tok = newToken(MINUS, l.ch)
	case '!': tok = newToken(BANG, l.ch)
	case '*': tok = newToken(ASTERISK, l.ch)
	case '/':
		if l.peekChar() == '/' {
			l.readChar(); l.readChar()
			tok.Type = COMMENT
			tok.Literal = l.readLineComment()
			return tok
		} else if l.peekChar() == '*' {
			l.readChar(); l.readChar()
			tok.Type = COMMENT
			tok.Literal = l.readBlockComment()
			return tok
		}
		tok = newToken(SLASH, l.ch)
	case '<': tok = newToken(LT, l.ch)
	case '>': tok = newToken(GT, l.ch)
	case '?': tok = newToken(QUESTION, l.ch)
	case '{': tok = newToken(LBRACE, l.ch)
	case '}': tok = newToken(RBRACE, l.ch)
	case '[': tok = newToken(LBRACKET, l.ch)
	case ']': tok = newToken(RBRACKET, l.ch)
	case '(': tok = newToken(LPAREN, l.ch)
	case ')': tok = newToken(RPAREN, l.ch)
	case ',': tok = newToken(COMMA, l.ch)
	case '@': tok = newToken(AT, l.ch)
	case '.': tok = newToken(DOT, l.ch)
	case '#': tok = newToken(HASH, l.ch)
	case '&': tok = newToken(AMPERSAND, l.ch)
	case '"', '\'':
		tok.Type = STRING
		tok.Literal = l.readString(l.ch)
		l.readChar()
		return tok
	case 0:
		tok.Literal = ""
		tok.Type = EOF
	default:
		if isLetter(l.ch) {
			tok.Literal = l.readIdentifier()
			tok.Type = LookupIdent(tok.Literal) // Check for keywords
			return tok
		} else if isDigit(l.ch) {
			tok.Type = NUMBER
			tok.Literal = l.readNumber()
			return tok
		} else {
			tok = newToken(ILLEGAL, l.ch)
		}
	}

	l.readChar()
	return tok
}

func (l *Lexer) skipWhitespace() { for l.ch == ' ' || l.ch == '\t' || l.ch == '\n' || l.ch == '\r' { l.readChar() } }
func (l *Lexer) readIdentifier() string { position := l.position; for isIdentifierChar(l.ch) { l.readChar() }; return l.input[position:l.position] }
func (l *Lexer) readString(quote byte) string { position := l.position + 1; for { l.readChar(); if l.ch == quote || l.ch == 0 { break } }; return l.input[position:l.position] }
func (l *Lexer) readLineComment() string { position := l.position; for l.ch != '\n' && l.ch != 0 { l.readChar() }; return l.input[position:l.position] }
func (l *Lexer) readBlockComment() string { position := l.position; for !(l.ch == '*' && l.peekChar() == '/') && l.ch != 0 { l.readChar() }; if l.ch != 0 { l.readChar(); l.readChar() }; return l.input[position : l.position-2] }
func (l *Lexer) readNumber() string {
	position := l.position
	for isDigit(l.ch) {
		l.readChar()
	}
	if l.ch == '.' && isDigit(l.peekChar()) {
		l.readChar()
		for isDigit(l.ch) {
			l.readChar()
		}
	}
	return l.input[position:l.position]
}

func isLetter(ch byte) bool { return 'a' <= ch && ch <= 'z' || 'A' <= ch && ch <= 'Z' || ch == '_' }
func isDigit(ch byte) bool { return '0' <= ch && ch <= '9' }
func isIdentifierChar(ch byte) bool { return isLetter(ch) || isDigit(ch) || ch == '-' }
func newToken(tokenType TokenType, ch byte) Token { return Token{Type: tokenType, Literal: string(ch)} }

func (l *Lexer) Input() string {
	return l.input
}

func (l *Lexer) Position() int {
	return l.position
}

func (l *Lexer) SetReadPosition(pos int) {
	l.readPosition = pos
	l.readChar()
}

func (l *Lexer) GetReadPosition() int {
	return l.readPosition
}
